File: C++/maximum-length-of-repeated-subarray.cpp
// Time:  O(m * n)
// Space: O(min(m, n))

// dp solution (99 ms)
class Solution {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        if (A.size() < B.size()) { 
            return findLength(B, A);
        }
        int result = 0;
        vector<vector<int>> dp(2, vector<int>(B.size() + 1));
        for (int i = 0; i < A.size(); ++i) {
            for (int j = 0; j < B.size(); ++j) {
                if (A[i] == B[j]) {
                    dp[(i + 1) % 2][j + 1] = dp[i % 2][j] + 1;
                    result = max(result, dp[(i + 1) % 2][j + 1]);
                } else {
                    dp[(i + 1) % 2][j + 1] = 0;
                }
            }
        }
        return result;
    }
};


// Time:  O(m * n * log(min(m, n)))
// Space: O(min(m, n))
// Binary search + rolling hash solution (36 ms)
class Solution2 {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        if (A.size() > B.size()) {
            return findLength(B, A);
        }
        int left = 0, right = min(A.size(), B.size()) + 1;
        while (left < right) {
            const auto mid = left + (right-left) / 2;
            if (!check(mid, A, B)) {  // find the min idx such that check(idx) == false
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left - 1;
    }
    
private:
    bool check(const int guess, const vector<int>& A, const vector<int>& B) {
        unordered_map<uint64_t, vector<int>> hashes;
        int i = 0;
        for (const auto& hash_a : rolling_hashes(A, guess)) {
            hashes[hash_a].emplace_back(i++);
        }
        int j = 0;
        for (const auto& hash_b : rolling_hashes(B, guess)) {
            for (const auto& i : hashes[hash_b]) {
                if (equal(A.begin() + i, A.begin() + i + guess, B.begin() + j)) {
                    return true;
                }
            }
            ++j;
        }
        return false;
    }
            
    vector<uint64_t> rolling_hashes(const vector<int>& source, const int length) {
        static const uint64_t M = 1000000007;
        static const uint64_t p = 113; 
        static const uint64_t p_inv = pow(p, M - 2, M);
        vector<uint64_t> result(source.size() - length + 1);
        uint64_t hash = 0, power = 1;
        if (length == 0) {
            return result;
        }
        for (int i = 0; i < source.size(); ++i) {
            hash = (hash + source[i] * power) % M;
            if (i < length - 1) {
                power = (power * p) % M;
            } else {
                result[i - (length - 1)] = hash;
                hash = (hash - source[i - (length - 1)]) * p_inv % M;
            }
        }
        return result;
    }
    
    uint64_t pow(uint64_t a,uint64_t b, uint64_t m) {
        a %= m;
        uint64_t result = 1;
        while (b) {
            if (b & 1) {
                result = (result * a) % m;
            }
            a = (a * a) % m;
            b >>= 1;
        }
        return result;
    }
};

File: Python/maximum-length-of-repeated-subarray.py
# Time:  O(m * n)
# Space: O(min(m, n))

import collections


class Solution(object):
    def findLength(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        if len(A) < len(B): return self.findLength(B, A)
        result = 0
        dp = [[0] * (len(B)+1) for _ in xrange(2)]
        for i in xrange(len(A)):
            for j in xrange(len(B)):
                if A[i] == B[j]:
                    dp[(i+1)%2][j+1] = dp[i%2][j]+1
                else:
                    dp[(i+1)%2][j+1] = 0
            result = max(result, max(dp[(i+1)%2]))
        return result


# Time:  O(m * n * log(min(m, n)))
# Space: O(min(m, n))
# Binary search + rolling hash solution (226 ms)
class Solution2(object):
    def findLength(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        if len(A) > len(B): return self.findLength(B, A)
        M, p = 10**9+7, 113
        p_inv = pow(p, M-2, M)
        def check(guess):
            def rolling_hashes(source, length):
                if length == 0:
                    yield 0, 0
                    return

                val, power = 0, 1
                for i, x in enumerate(source):
                    val = (val + x*power) % M
                    if i < length - 1:
                        power = (power*p) % M
                    else:
                        yield val, i-(length-1)
                        val = (val-source[i-(length-1)])*p_inv % M

            hashes = collections.defaultdict(list)
            for hash_val, i in rolling_hashes(A, guess):
                hashes[hash_val].append(i)
            for hash_val, j in rolling_hashes(B, guess):
                if any(A[i:i+guess] == B[j:j+guess] for i in hashes[hash_val]):
                    return True
            return False

        left, right = 0, min(len(A), len(B)) + 1
        while left < right:
            mid = left + (right-left)/2
            if not check(mid):  # find the min idx such that check(idx) == false
                right = mid
            else:
                left = mid+1
        return left-1


# Time:  O(m * n * min(m, n) * log(min(m, n)))
# Space: O(min(m^2, n^2))
# Binary search (122 ms)
class Solution3(object):
    def findLength(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        if len(A) > len(B): return self.findLength(B, A)

        def check(length):
            lookup = set(A[i:i+length] \
                       for i in xrange(len(A)-length+1))
            return any(B[j:j+length] in lookup \
                       for j in xrange(len(B)-length+1))

        A = ''.join(map(chr, A))
        B = ''.join(map(chr, B))
        left, right = 0, min(len(A), len(B)) + 1
        while left < right:
            mid = left + (right-left)/2
            if not check(mid):  # find the min idx such that check(idx) == false
                right = mid
            else:
                left = mid+1
        return left-1

